t/protocol_directnic.pl

```diff
@@ -0,0 +1,231 @@
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+            {label => 'SUCCESS', ctx => ['h2'], msg => qr/IPv4/},
+        ],
+    },
+    {
+        desc => 'IPv6, single host, good',
+        cfg => {h1 => {url => "$hostname/dns/gateway/abc/"}},
+        wantips => {h1 => {wantipv6 => '2001:db8::1'}},
+        resp => ['{"result":"success","message":"Your record was updated to 2001:db8::1"}'],
+        wantreq => '/dns/gateway/abc/?data=2001:db8::1',
+        wantstatus => {
+            h1 => {'status-ipv6' => 'good', 'ipv6' => '2001:db8::1', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, good',
+        cfg => {h1 => {urlv4 => "$hostname/dns/gateway/abc/", urlv6 => "$hostname/dns/gateway/def/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1', wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.1',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv4/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, mixed success',
+        cfg => {h1 => {urlv4 => "$hostname/dns/gateway/bad_token/", urlv6 => "$hostname/dns/gateway/def/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1', wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'failed',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+);
+
+for my $tc (@test_cases) {
+    diag('==============================================================================');
+    diag("Starting test: $tc->{desc}");
+    diag('==============================================================================');
+    local $ddclient::globals{debug} = 1;
+    local $ddclient::globals{verbose} = 1;
+    my $l = Logger->new($ddclient::_l);
+    local %ddclient::config;
+    $ddclient::config{$_} = {
+        protocol => 'directnic',
+        %{$tc->{cfg}{$_}},
+    } for keys(%{$tc->{cfg}});
+    # Run init_config(), this is what loads JSON support based on the protocols defined in the config
+    ddclient::init_config();
+    # init_config() will filter out config properties, but the part of update_nics() that populates wantip*
+    # is hard to extract, so we'll just wire in our desired wantip* config directly here
```

Comment:
I would just call `ddclient::load_json_support();` at the top of this file.  That wouldn't test proper loading of JSON support, but this is a unit test, not an integration test, and it's easier to reason about what the test is doing if `init_config` is not called.

Reply:
That does make this a lot cleaner. That's also the reason I'd changed the status key mask below, because `init_config` initializes some of those variables to `undef` while they were not present in `wantstatus`. Glad to be able to revert that to the example from `t/protocol_dyndns2.pl`.

This does mean my changes to add `T_URL` as a supported type aren't exercised as part of this test, but I agree those would be better off in another unit test that targets those specifically. I added `t/check_values.pl`, following the pattern of `t/variable_defaults.pl`. I welcome any feedback on that test.